### ISSUE TITLE
iterativelySimplifyCFG: Assertion `IterCnt++ < 1000 && "Iterative simplification didn't converge!"' failed.

### ISSUE THREAD

**Issue by JonPsson** [llvm/llvm-project#54610]
[testcase.tar.gz](https://github.com/llvm/llvm-project/files/8369253/testcase.tar.gz)

opt -mtriple=s390x-linux-gnu -mcpu=z15 -O3 ./tc_crash1_aftercreduce.ll -o /dev/null -dse-memoryssa-defs-per-block-limit=200

SimplifyCFGPass.cpp:240: bool iterativelySimplifyCFG(llvm::Function&, const llvm::TargetTransformInfo&, llvm::DomTreeUpdater*, const llvm::SimplifyCFGOptions&): Assertion `IterCnt++ < 1000 && "Iterative simplification didn't converge!"' failed.

**Comment by fhahn** [comment]
@JonPsson do you know if that's a recent regression?

**Comment by max-quazan** [comment]
This is an old issue, one possible explanation was provided here: https://reviews.llvm.org/D112693

This fix wasn't accepted and I have no alternatives.

**Comment by max-quazan** [comment]
Some context: the hang has existed for few years now. I've added this assertion to make sure it's an assertion failure and not hang at least in fastdebug. So it's only highlighting the problem that we found back in October.

**Comment by fhahn** [comment]
Got it, thanks @max-quazan ! So this is related/ likely a duplicate of #51632 IIUC

**Comment by nikic** [comment]
I don't think this is the same issue. From `-print-after-all` output there is some massive unrolling going on here, and SimplifyCFG needs a lot of iterations to simplify that.

I haven't looked into what causes the extreme unrolling here.
